### DISCUSSION TITLE
Numerical differences between two setups of reversible reaction in MoBi

### DISCUSSION THREAD

**Discussion by moPellowe** [Open-Systems-Pharmacology/Forum#1176] [Questions & Problems] [answered]
I've set up a reversible reaction in MoBi two different ways that I had expected to be equivalent: 1) as a single reaction and 2) as two separate reactions (one for binding and the other for disassociation). The parameters used result in numerical instability, but it seems that using separate reactions allows the simulation to run at a higher tolerance. Why would this be the case?
![image](https://user-images.githubusercontent.com/57540473/158847276-35c1e56b-bacb-4bbc-9ee6-212ff8e95790.png)
![image](https://user-images.githubusercontent.com/57540473/158847498-f9f89036-fe32-4e13-b994-55c9af075d25.png)

Both simulations will run for 24 hours, but only the 2nd one will run for 2 years (with tolerances = 1E-10). I can also lower the tolerances for the single reaction simulation to 1E-16 so that it also runs, but then the results of the two simulations are different. The simulation with separate reactions is very noisy but it can run at a tolerance of 1E-10. The single reaction needs the smaller tolerance of 1E-16 but it looks cleaner. So is the single reaction simulation more correct?

![image](https://user-images.githubusercontent.com/57540473/158848243-f7eab1b5-47a3-495c-9d54-48a32a544ea6.png)
![image](https://user-images.githubusercontent.com/57540473/158848189-f40f7ff2-1f3c-4b92-9cbf-c9ad70fe7065.png)

I've attached the project file below. 
[test.zip](https://github.com/Open-Systems-Pharmacology/Forum/files/8288061/test.zip)

**Comment by msevestre** [comment]
Can you try using the scaling factor features and see if it makes any difference?

**Reply by moPellowe** [reply]
I haven't used the scaling factor features before. Where do I find that?

**Reply by moPellowe** [reply]
I tried using the "Calculate Scale Divisor" feature (thank you @Christoph27 for your help), and that helped with the tolerances but only partially with the simulation results. With the scaling factors, both simulations are able to run with tolerances of 1E-10, and the separate simulation is not noisy anymore. However, the results are still different from each other, which is likely due to the different ODEs @msevestre showed below. From the looks of the equations, I guess the single reaction is correct since it uses the same value for k_off instead of having a separate k_off value for the dissasociation reaction. Does that sounds right?

**Reply by msevestre** [reply]
It looks to me like your systems are not equivalent. It looks like some signs are even opposite (+ vs -). Can it be that some educts are defined as products or sthg like this?

**Reply by moPellowe** [reply]
Do you mean that the molecule type of the educt(s) matters? Because in the simulation, A and B are type Drug while AB is type Complex.

**Comment by PavelBal** [comment]
I am very interested in the answer, too. I thought both ways would result in exact the same ODE system.

**Comment by msevestre** [comment]
Can you generate the ode and see the difference? (Start using the --dev mode).

**Reply by moPellowe** [reply]
I generated simulation reports to compare the two, and I did notice that the k_off value is different for the dissasociation reaction compared to the single reaction. Could this account for the difference?

![image](https://user-images.githubusercontent.com/57540473/158856972-57bdd52c-e815-4d8f-bf5c-1fb15ccb1012.png)

I do think I set things up correctly since the value for k_off = 1/h for both simulations.

![image](https://user-images.githubusercontent.com/57540473/158857768-509cf777-9b61-4342-a384-1807dd540542.png)
![image](https://user-images.githubusercontent.com/57540473/158858171-b98c53db-b1d0-4c00-8c4a-cb2b5f00e1ad.png)

**Comment by Christoph27** [comment]
You can calculate the scale divisors using the button marked in the attached figure (if you are in a simulation). The scale divisors become part of the MSV building block after calculation (see the column there). Highly recommended if some of your state variables have very small values.
![ScaleDivisor](https://user-images.githubusercontent.com/26439150/158856282-9b610284-37c3-46c0-a1e8-7eae9def1a2c.jpg)

**Reply by moPellowe** [reply]
I'll give that a try. Thanks.

**Comment by msevestre** [comment]
Heare the matlab output 

```
function dy = ODERHSFunction(Time,y)

P_0 = 1;  % Test|Organism|Volume
P_1 = 6;  % Test|Organism|Rxn|k_on
P_2 = 24;  % Test|Organism|Rxn|k_off
P_3 = 1e-16;  % Test|Organism|Rxn|V
P_4 = 1e-16;  % AbsTol
P_5 = 1e-16;  % RelTol
P_6 = 1e-10;  % H0
P_7 = 0;  % HMin
P_8 = 60;  % HMax
P_9 = 100000;  % MxStep
P_10 = 1;  % UseJacobian
P_11 = (IIf((P_0 > 0),(y(1)/P_0),0));  % Test|Organism|A|Concentration
P_12 = (IIf((P_0 > 0),(y(2)/P_0),0));  % Test|Organism|B|Concentration
P_13 = (IIf((P_0 > 0),(y(3)/P_0),0));  % Test|Organism|AB|Concentration

dy(1) =  ...
            (0-((P_1*y(1)*(y(2)/P_3))-(P_2*y(3))));
dy(2) =  ...
            (0-((P_1*y(1)*(y(2)/P_3))-(P_2*y(3))));
dy(3) =  ...
            ((P_1*y(1)*(y(2)/P_3))-(P_2*y(3)));
dy = dy';

```

vs 

```
function dy = ODERHSFunction(Time,y)

P_0 = 1;  % separate|Organism|Volume
P_1 = 6;  % separate|Organism|Rxn|k_on
P_2 = 24;  % separate|Organism|Rxn|k_off
P_3 = 1e-16;  % separate|Organism|Rxn|V
P_4 = 0.0166666666666667;  % separate|Organism|Disassociate|k_off
P_5 = 1e-10;  % AbsTol
P_6 = 1e-10;  % RelTol
P_7 = 1e-10;  % H0
P_8 = 0;  % HMin
P_9 = 60;  % HMax
P_10 = 100000;  % MxStep
P_11 = 1;  % UseJacobian
P_12 = (IIf((P_0 > 0),(y(1)/P_0),0));  % separate|Organism|A|Concentration
P_13 = (IIf((P_0 > 0),(y(2)/P_0),0));  % separate|Organism|B|Concentration
P_14 = (IIf((P_0 > 0),(y(3)/P_0),0));  % separate|Organism|AB|Concentration

dy(1) =  ...
            (0-(P_1*y(1)*(y(2)/P_3))) ...
            +(P_4*y(3));
dy(2) =  ...
            (0-(P_1*y(1)*(y(2)/P_3))) ...
            +(P_4*y(3));
dy(3) =  ...
            (P_1*y(1)*(y(2)/P_3)) ...
            +(0-(P_4*y(3)));
dy = dy';
```

**Reply by msevestre** [reply]
Those don't look equivalent to me....

**Comment by PavelBal** [comment, marked as answer]
The values of used `k_off` are different..

**Reply by moPellowe** [reply]
Oh! Is it just a typo then in the simulation?

**Reply by moPellowe** [reply]
Oh, yes, I see that now. Thank you for pointing that out. Sorry for the trouble. 😅

**Reply by moPellowe** [reply]
So I went back to my main model file and I'm still getting different results when (I believe) the only difference is the Reaction building block: one having two reactions for each reversible reaction and the other combining them into one. I'm having a hard time tracking down the typo in the simulations and/or the Reaction building block. Any help would be greatly appreciated.

I've attached (now removed) a file with a single simulation. When I swap the Reaction building block, I get different results.

**Reply by Christoph27** [reply]
In the building block "Reaction - separate" the "Container Criteria" for the binding and dissociation reaction are not the same. Could it be that they are not happening in parallel in the same compartments as they should?

**Reply by moPellowe** [reply]
Oh, thank you so much for finding that! I'll give that a try.
